File: src/components/sdl3/fio.lua
---@meta fio

--- @enum Jila_IO_Whence
Jila_IO_Whence = {
    SET = 0,
    CUR = 1,
    END = 2
}

--- @class Jila_IO
Jila_IO = {}

--- Writes data to the stream.
--- @param stream Jila_IO The stream to write to.
--- @param data string The string data to write.
--- @return number number The number of bytes written.
function Jila_IO_Write(stream, data) end

--- Reads data from the stream.
--- @param stream Jila_IO The stream to read from.
--- @param size number The number of bytes to read.
--- @return string string The data read.
function Jila_IO_Read(stream, size) end

--- Seeks to an offset in the stream.
--- @param stream Jila_IO The stream to seek in.
--- @param offset number The offset to seek to.
--- @param whence Jila_IO_Whence The reference point for the offset.
--- @return number The new position in the stream, or -1 on error.
--- 
--- @see Jila_GetError for read error.
function Jila_IO_Seek(stream, offset, whence) end

--- Gets the current position in the stream.
--- @param stream Jila_IO The stream.
--- @return number The current position, or -1 on error.
--- 
--- @see Jila_GetError for read error.
function Jila_IO_Tell(stream) end

--- Gets the size of the stream.
--- @param stream Jila_IO The stream.
--- @return number The size of the stream, or -1 on error.
--- 
--- @see Jila_GetError for read error.
function Jila_IO_GetSize(stream) end

--- Opens a file as an IO stream.
--- @param path string The path to the file.
--- @param mode string The mode to open the file in (e.g., "rb", "wb").
--- @return Jila_IO|nil Jila_IO The IO stream, or nil on error.
--- 
--- @see Jila_GetError for read error.
function Jila_IO_Open(path, mode) end

File: src/components/sdl3/fio.cpp
#include "components/sdl3/fio.hpp"
#include "misc.hpp"

namespace Jila {

IOStream _SDL_IO_Open(
    std::string_view path,
    std::string_view mode
) {
    SDL_IOStream* s = SDL_IOFromFile(
        path.data(), mode.data()
    );

    return MakeSafeMemory<SDL_IOStream>(
        s,
        [](SDL_IOStream* s) {
            if (s) SDL_CloseIO(s);
        }
    );
}

size_t _SDL_IO_Write(
    IOStream stream,
    std::string_view data
) {
    return SDL_WriteIO(stream.get(), data.data(), data.size());
}

std::string _SDL_IO_Read(IOStream stream, size_t size) {
    std::string buffer(size, '\0');
    size_t nr = SDL_ReadIO(stream.get(), buffer.data(), size);
    buffer.resize(nr);
    return buffer;
}

long long _SDL_IO_Seek(
    IOStream stream, 
    long long offset, 
    int whence
) {
    return SDL_SeekIO(stream.get(), offset, (SDL_IOWhence)whence);
}

long long _SDL_IO_Tell(IOStream stream) {
    return SDL_TellIO(stream.get());
}

long long _SDL_IO_GetSize(IOStream stream) {
    return SDL_GetIOSize(stream.get());
}

void bindSdlFio(sol::state* state) {
    state->set_function("Jila_IO_Open", &_SDL_IO_Open);
    state->set_function("Jila_IO_Write", &_SDL_IO_Write);
    state->set_function("Jila_IO_Read", &_SDL_IO_Read);
    state->set_function("Jila_IO_Seek", &_SDL_IO_Seek);
    state->set_function("Jila_IO_Tell", &_SDL_IO_Tell);
    state->set_function("Jila_IO_GetSize", &_SDL_IO_GetSize);

    state->new_enum("Jila_IO_Whence",
        "SET", SDL_IO_SEEK_SET,
        "CUR", SDL_IO_SEEK_CUR,
        "END", SDL_IO_SEEK_END
    );
}

}
